### PR TITLE
Call the Shopify on the Solidus Update / Destroy / Create for a product

### PR DESCRIPTION
**Pull request by thisiscab**
Problematic case:

Once a Solidus gets updated, we want to trigger an after_update event
that will trigger the task that will update this product to Shopify.

When we update it on Shopify, in a certain case the product on Shopify
may have been deleted so we need to re-create it. When we re-create it,
we need to save the shopify_product_id to the Solidus product. And there
goes an infinite loop

This solves that.

### CODE REVIEW

File: app/models/spree/product_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    def create_shopify_product
+      service = Shopify::ProductExporter.new(spree_product: self)
+      service.perform
+    end
+
+    def update_shopify_product
+      if assembly?
+        service = Shopify::BundledProductExporter.new(spree_product: self)
+      else
+        service = Shopify::ProductUpdater.new(spree_product_id: id)
```
Comment (thisiscab) [inline app/models/spree/product_decorator.rb:22]: We probably have to change the initializer interface here

File: app/models/spree/product_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    private
+
+    def create_shopify_product
+      service = Shopify::ProductExporter.new(spree_product: self)
+      service.perform
+    end
+
+    def update_shopify_product
+      if assembly?
```
Comment (thisiscab) [inline app/models/spree/product_decorator.rb:19]: I am unsure if this should be done in some kind of factories, what do you guys think?

**Comment by thisiscab** [comment]
Whenever you guys are ready! 
@DanielWright @braidn @bryanmtl 💃

**Comment by braidn** [comment]
This is much more manageable. Thank you.

**Comment by thisiscab** [comment]
Damnnn, that implementation can't work. Since I'm creating Solidus objects in my specs to test this feature. So when I create a Solidus object in my specs, it directly exports them to Shopify, which makes all kind of funky results.

Do you guys have suggestions?

I could go in the admin controller, decorate it so that it does it only there...

**Comment by DanielWright** [comment]
Can you make callbacks conditional? Maybe implement a predicate method (e.g. `should_export_to_shopify?`) that encapsulates the conditions under which a product should or should not be sent to Shopify?

**Comment by thisiscab** [comment]
That would have been a good idea, but I can't think of a condition where we wouldn't want to save it to Shopify :/

We could have a flag that is turned on or off depending on some scenario, but that seems funky to me.

**Comment by DanielWright** [comment]
If the situation is you don't want _your tests_ to export your products to Shopify, then why not adjust your test environment to stub that behaviour? A tool like WebMock would allow you to stub requests to Shopify such that your object think it's talking to Shopify, but the messages are going nowhere.

File: app/models/spree/product_decorator.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+module Spree
+  module ProductDecorator
```
Comment (braidn) [inline app/models/spree/product_decorator.rb:2]: Is this something we want to decorate onto the Product class? Or should we build some kind of factory class that encapsulates all the magic?

File: app/services/shopify/association_saver.rb [hunk 4]
```diff
@@ -11,16 +11,14 @@ def save_pos_variant_id_for_variants(spree_variants, shopify_variants)
       end
 
       def save_pos_variant_id(spree_variant, shopify_variant)
-        spree_variant.pos_variant_id = shopify_variant.id
-        spree_variant.save
+        spree_variant.update_columns(pos_variant_id: shopify_variant.id)
```
Comment (braidn) [inline app/services/shopify/association_saver.rb:14]: I know there was a question about how this doesn't initiate callbacks this morning. However, I read these methods like:

`we are updating the column directly in the db and bypassing the model completely`

This is why I think it reads fine and clues the reader into what's 'sort of' going on.

File: spec/models/spree/product_spec.rb [hunk 5]
```diff
@@ -0,0 +1,88 @@
+require 'spec_helper'
+
+module Spree
+  RSpec.describe Product do
```
Comment (braidn) [inline spec/models/spree/product_spec.rb:4]: Yeah reading the decorator and this, my twitch would be to create a class/object to encapsulate this work that's not the product itself. This also helps with removing God objects which tend to rear their heads on User/Product/Order. Totally up to you though.

**Review by braidn — COMMENTED** [review]
MUCH better Cab. Easliy able to look through and lay 👀 on the all of the code. Thank you again.

**Comment by thisiscab** [comment]
I'd like to touch base about the feature that this adds here. I am hitting so many problems with that implementation, where it sends the data to shopify on callbacks. Having a verbal discussion would help me :)

**Comment by braidn** [comment]
Write away! Or... did you mean phone/video?

**Comment by thisiscab** [comment]
Phone

**Comment by thisiscab** [comment]
I would like to continue the conversation here @braidn https://github.com/glossier/solidus_retail/pull/34/files

**Comment by thisiscab** [comment]
We have hit many problems with the way callback works. It was really hard to make sure that under certain scenario stuff wouldn't save to Shopify but in other they would. 

Instead we have decided to implement those calls directly in the action of the Admin Controller.

**Comment by thisiscab** [comment]
We have hit many problems with the way callback works. It was really hard to make sure that under certain scenario stuff wouldn't save to Shopify but in other they would. 

Instead we have decided to implement those calls directly in the action of the Admin Controller.

**Pull request closed without merging**